core/vm/interpreter.go

```diff
@@ -918,3 +918,159 @@ func (in *GethEVMInterpreter) runPlain(state *InterpreterState, input []byte, re
+	if len(state.Contract.Code) == 0 {
+		return
+	}
+
+	// Don't execute if already halted.
+	if state.Halted {
+		return
+	}
+
+	// Get the operation from the jump table and validate the stack to ensure there are
+	// enough stack items available to perform the operation.
+	state.op = state.Contract.GetOp(state.Pc)
+	operation := in.cfg.JumpTable[state.op]
+	if operation == nil {
+		state.Err = &ErrInvalidOpCode{opcode: state.op}
+		return
+	}
+	// Validate stack
+	if sLen := state.Stack.len(); sLen < operation.minStack {
+		state.Err = &ErrStackUnderflow{stackLen: sLen, required: operation.minStack}
+		return
+	} else if sLen > operation.maxStack {
+		state.Err = &ErrStackOverflow{stackLen: sLen, limit: operation.maxStack}
+		return
+	}
+	// If the operation is valid, enforce write restrictions
+	if in.readOnly && in.evm.chainRules.IsByzantium {
+		// If the interpreter is operating in readonly mode, make sure no
+		// state-modifying operation is performed. The 3rd stack item
+		// for a call operation is the value. Transferring value from one
+		// account to the others means the state is modified and should also
+		// return with an error.
+		if operation.writes || (state.op == CALL && state.Stack.Back(2).Sign() != 0) {
+			state.Err = ErrWriteProtection
+			return
+		}
+	}
+	// Static portion of gas
+	if !state.Contract.UseGas(operation.constantGas) {
+		state.Err = ErrOutOfGas
+		return
+	}
+
+	var memorySize uint64
+	// calculate the new memory size and expand the memory to fit
+	// the operation
+	// Memory check needs to be done prior to evaluating the dynamic gas portion,
+	// to detect calculation overflows
+	if operation.memorySize != nil {
+		memSize, overflow := operation.memorySize(state.Stack)
+		if overflow {
+			state.Err = ErrGasUintOverflow
+			return
+		}
+		// memory is expanded in words of 32 bytes. Gas
+		// is also calculated in words.
+		if memorySize, overflow = math.SafeMul(toWordSize(memSize), 32); overflow {
+			state.Err = ErrGasUintOverflow
+			return
+		}
+	}
+	// Dynamic portion of gas
+	// consume the gas and return an error if not enough gas is available.
+	// cost is explicitly set so that the capture state defer method can get the proper cost
+	if operation.dynamicGas != nil {
+		dynamicCost, err := operation.dynamicGas(in.evm, state.Contract, state.Stack, state.Memory, memorySize)
+		if err != nil || !state.Contract.UseGas(dynamicCost) {
+			state.Err = ErrOutOfGas
+			return
+		}
+	}
+	if memorySize > 0 {
+		state.Memory.Resize(memorySize)
+	}
+
+	state.Result, state.Err = operation.execute(&state.Pc, in, state.CallContext)
+
+	// if the operation clears the return data (e.g. it has returning data)
+	// set the last return to the result of the operation.
+	if operation.returns {
+		in.returnData = state.Result
+	}
+
+	switch {
+	case operation.reverts:
+		state.Err = ErrExecutionReverted
+	case operation.halts:
+		state.Halted = true
+	case !operation.jumps:
+		state.Pc++
+	}
```

Comment:
I think this function covers many of the aspects under test and should thus share the same code as the actual Geth implementation of line 834-917. It would be best to factor out this part into a `step` function and call it from both entry points. Is this blocked by something?

Reply:
That is a very good point, and I think that should be possible. However, this would still leave the implementations in `runBasicBlockProfiling` and `runMicroProfiling` as-is, which means those would not share the same code. So there could still be bugs hiding in this code duplication.